atomic_reactor/plugins/pre_fetch_sources.py

```diff
@@ -0,0 +1,226 @@
+        """Fetch SRPM download URLs for each image generated by a build
+
+        Build each possible SRPM URL and check if the URL is available,
+        respecting the signing intent preference order.
+
+        :param build_id: int, Koji build id of the container image we want SRPMs for
+        :param sigkeys: list, strings for keys which signed the srpms to be fetched
+        :return: list, strings with URLs pointing to SRPM files
+        """
+        if not sigkeys:
+            sigkeys = ['']
+
+        srpm_build_paths = {}
+        archives = self.session.listArchives(build_id, type='image')
+        for archive in archives:
+            rpms = self.session.listRPMs(imageID=archive['id'])
+            for rpm in rpms:
+                rpm_hdr = self.session.getRPMHeaders(rpm['id'], headers=['SOURCERPM'])
+                srpm_filename = rpm_hdr['SOURCERPM']
+                if srpm_filename in srpm_build_paths:
+                    continue
+                rpm_build = self.session.getBuild(rpm['build_id'], strict=True)
+                path_info = get_koji_path_info(self.workflow, self.koji_fallback)
+                base_url = path_info.build(rpm_build)
+                srpm_build_paths[srpm_filename] = base_url
+
+        srpm_urls = []
+        missing_srpms = []
+        for srpm_filename, base_url in srpm_build_paths.items():
+            srpm_url = None
+            for sigkey in sigkeys:
+                # koji uses lowercase for paths. We make sure the sigkey is in lower case
+                url_candidate = self.assemble_srpm_url(base_url, srpm_filename, sigkey.lower())
+                req_session = get_retrying_requests_session()
+                request = req_session.head(url_candidate, verify=not insecure)
+                if request.ok:
+                    srpm_url = url_candidate
+                    break
+                self.log.debug('%s not found for signing key "%s" at %s (returned %s)',
+                               srpm_filename, sigkey, url_candidate, request.status_code)
+
+            if not srpm_url:
+                missing_srpms.append(srpm_filename)
+            else:
+                srpm_urls.append(srpm_url)
+
+        if missing_srpms:
+            raise RuntimeError('Could not find files signed by any of {} for these SRPMS: {}'
+                               .format(sigkeys, srpm_filename))
+
+        return srpm_urls
+
+    def get_signing_intent(self, build_id):
+        """Get the signing intent to be used to fetch files from Koji
+
+        :param build_id: int, Koji build id of the container image which owns the signing intent
+        :return: dict, signing intent object as per atomic_reactor/schemas/config.json
+        """
+        odcs_config = get_config(self.workflow).get_odcs_config()
+        if not self.signing_intent:
+            if not self.koji_build:
+                self.koji_build = self.session.getBuild(build_id, strict=True)
```

Comment:
Will this line ever get called?

Reply:
If an NVR for the image is not passed(let's say just a build id was provided), getBuild will not be called in `get_image_build_id`. Then if a signing intent from koji is needed, this will be called.